internal/wasi/fs.go

```diff
@@ -5,107 +5,317 @@ import (
 	"io"
 	"io/fs"
 	"os"
+	"path"
 	"runtime"
+	"sort"
 	"strings"
-
-	"github.com/tetratelabs/wazero/wasi"
+	"time"
 )
 
-type DirFS string
+// DirFS is a file system that implements fs.FS and wasi.OpenFileFS, similar to the read-only os.DirFS.
+// This is the only implementation that allows the Wasm to create or change files on the host file system, since
+// there is no official alternative interface to wasi.OpenFileFS.
+// See os.DirFS.
+type DirFS string // string holds the root path of this directory on the host file system.
 
-func (dir DirFS) OpenWASI(dirFlags uint32, path string, oFlags uint32, fsRights, fsRightsInheriting uint64, fdFlags uint32) (wasi.File, error) {
-	// I'm not sure how to use all these passed flags and rights yet
-	if !fs.ValidPath(path) || runtime.GOOS == "windows" && strings.ContainsAny(path, `\:`) {
-		return nil, &os.PathError{Op: "open", Path: path, Err: os.ErrInvalid}
-	}
-	mode := fs.FileMode(0644)
-	if oFlags&wasi.O_DIR != 0 {
-		mode |= fs.ModeDir
+// OpenFile implements wasi.OpenFileFS.OpenFile.
+func (dir DirFS) OpenFile(name string, flag int, perm os.FileMode) (fs.File, error) {
+	// fs.FS requires a FS to reject a path that doesn't satisfy fs.ValidPath.
+	if !fs.ValidPath(name) ||
+		// '\' works as alternative path separater and ':' allows to express a root drive directory.
+		// fs.FS implementation must reject those path on Windows. See Note in the doc of fs.ValidPath.
```

Comment:
this is a good example of a leaky abstraction. WASI isn't windows, and there's nothing about fs.FS that should ever allow windows paths to leak into it. Ex. if someone exposes their "c:" drive, that's not on us to enforce. The only sensible file mappings would be chrooted if even an actual filesystem.

Reply:
In other words windows-based file handing in a barely specified posix system layer is a non-goal. We can use this to simplify discussions in future work. By adding code to address things like OS-specific things inside WASI, we are actually making the problem more confused, especially to people new who may not know Wasm is effectively a virtual machine, and if shares anything in common with OS, it would be unix, not windows, that we'd optimize for.